src/org/opensolaris/opengrok/analysis/StreamSource.java

```diff
@@ -62,6 +79,58 @@ public static StreamSource fromFile(final File file) {
             public InputStream getStream() throws IOException {
                 return new BufferedInputStream(new FileInputStream(file));
             }
+
+            @Override
+            public DigestedInputStream getSHA256stream() throws IOException {
+                return getSHA256stream(getStream());
+            }
         };
     }
+
+    protected static DigestedInputStream getSHA256stream(InputStream under)
+            throws IOException {
+        MessageDigest md;
+        try {
+            md = MessageDigest.getInstance(SHA256);
+        } catch (NoSuchAlgorithmException ex) {
+            /*
+             * NOT EXPECTED because "Every implementation of the Java
+             * platform is required to support ... SHA-256 ...."
+             */
+            throw new IOException("digest failed");
+        }
+        DigestInputStream dis = new DigestInputStream(under, md);
+
+        return new DigestedInputStream() {
+            @Override
+            public InputStream getStream() {
+                return dis;
+            }
+
+            @Override
+            public MessageDigest getMessageDigest() {
+                return dis.getMessageDigest();
+            }
+
+            @Override
+            public byte[] digestAll() throws IOException {
+                return StreamSource.digestAll(this);
+            }
+
+            @Override
+            public void close() throws Exception {
+                dis.close();
+            }
+        };
+    }
+
+    protected static byte[] digestAll(DigestedInputStream dis)
+            throws IOException {
+        byte[] buf = new byte[64 * 1024];
+        InputStream iss = dis.getStream();
+        while (iss.read(buf) != -1) {
+            // iterate
```

Comment:
let's see what will be the performance impact,
but since we traverse the same file maybe like for the 4th time, I guess it'll be OK

Reply:
[README](https://github.com/oracle/opengrok/blob/master/README.md) does advise that "Source base should be available locally for OpenGrok to work efficiently"; so maybe we count on ARC or the Linux equivalent to let additional reads be minimally more costly.

Reply:
The ZFS ARC can be big beast no doubt however what I see in the production (on what can be considered now a modest machine with 256 GB of RAM) is that JVM memory pressure forces ARC to shrink during indexing quite drastically, especially with parallel project indexing. Granted, I could tune `zfs_arc_min` to avoid that however that would hamper the indexer(s). You can call it bad capacity planning :-)

Reply:
In this case I just mean ARC enough for a sequence of file reads (sequential in OpenGrok; parallel in Lucene) during any single `addFile()` call. I.e., not long-term ARC for these files.

Reply:
Still, could you maybe do some quick benchmark (say with thousands of files changed) to see if the performance impact is indeed negligible ?

Reply:
It's imperceptible. The best wipe/reindex of three for `feature/doc_sha` (f0ad75794):

```
2018-02-20 08:32:43.462-0600 FINE t40 PendingFileCompleter.complete: renamed 1,329 file(s)
2018-02-20 08:32:43.462-0600 FINE t40 IndexDatabase.finishWriting: completed 1,329 file(s)
2018-02-20 08:32:43.464-0600 INFO t40 IndexDatabase.optimize: Optimizing the index ... 
2018-02-20 08:32:44.572-0600 INFO t40 IndexDatabase.optimize: done
...
2018-02-20 08:32:44.583-0600 INFO t1 Statistics.report: Done ... (took 27.869 seconds)
```

... was a quarter-second faster than for `master` (efc7f934900):

```
2018-02-20 08:29:47.553-0600 FINE t40 PendingFileCompleter.complete: renamed 1,329 file(s)
2018-02-20 08:29:47.553-0600 FINE t40 IndexDatabase.finishWriting: completed 1,329 file(s)
2018-02-20 08:29:47.555-0600 INFO t40 IndexDatabase.optimize: Optimizing the index ... 
2018-02-20 08:29:48.882-0600 INFO t40 IndexDatabase.optimize: done
...
2018-02-20 08:29:48.896-0600 INFO t1 Statistics.report: Done ... (took 28.173 seconds)
```